### PR TITLE
jiten: init at 1.0.0

### PR DESCRIPTION
**Pull request by obfusk**
###### Motivation for this change

I'm the developer of [Jiten Japanese Dictionary](https://github.com/obfusk/jiten). A NixOS user wanted to install it, so I made a nix expression. Since having it in nixpkgs would be even better, I'm making this PR :)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] ~~Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))~~
- [ ] ~~Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`~~
- [ ] ~~Tested execution of all binary files (usually in `./result/bin/`)~~
- [ ] ~~Determined the impact on package closure size (by running `nix path-info -S` before and after)~~
- [ ] ~~Ensured that relevant documentation is up to date~~
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/misc/jiten/default.nix [hunk 1]
```diff
@@ -0,0 +1,79 @@
+{ lib
+, fetchFromGitHub
+, python3Packages
+, makeWrapper
+, pcre
+, sqlite
+, python3
+, rlwrap
+}:
+python3Packages.buildPythonApplication rec {
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/jiten/default.nix:10]:
```suggestion
}:

python3Packages.buildPythonApplication rec {
```

File: pkgs/applications/misc/jiten/default.nix [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  pname                 = "jiten";
+  version               = "0.4.0";
+  buildInputs           = [ makeWrapper pcre sqlite ];
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/jiten/default.nix:13]:
```suggestion
  pname = "jiten";
  version = "0.4.0";

  src = fetchFromGitHub {
    owner = "obfusk";
    repo = "jiten";
    rev = "v${version}";
    sha256  = "0l690lfj5rlh9y6aqysh1sgwh13f2yq0vsdq4dimxxpjqixfk20x";
  };

  nativeBuildInputs = [ makeWrapper ];
  buildInputs = [ pcre sqlite ];
```

File: pkgs/applications/misc/jiten/default.nix [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  JITEN_VERSION         = version;  # override `git describe`
+  JITEN_ANDROID         = "yes";    # build & package *.sqlite3
+  patchPhase = ''
+    patchShebangs bin/jiten
+    sed -i 's!/bin/bash!'"$(command -v bash)"'!' Makefile jiten/res/jmdict/Makefile
+  '';
+  postFixup = ''
+    wrapProgram $out/bin/jiten --prefix PATH : ${lib.makeBinPath [ python3 rlwrap ]} \
+                               --prefix PYTHONPATH : "$program_PYTHONPATH"
+  '';
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/jiten/default.nix:24]:
```suggestion
  propagatedBuildInputs = with python3Packages; [ click flask python3 rlwrap ];

  JITEN_VERSION = version;  # override `git describe`
  JITEN_ANDROID = "yes";    # build & package *.sqlite3

  postPatch = ''
    patchShebangs bin/jiten
    sed -i 's!/bin/bash!'"$(command -v bash)"'!' Makefile jiten/res/jmdict/Makefile
  '';

  postFixup = ''
    wrapProgram $out/bin/jiten \
      --prefix PATH : ${lib.makeBinPath [ python3 rlwrap ]} \
      --prefix PYTHONPATH : "$program_PYTHONPATH"
  '';

```

Can we export the variables in preBuild?

File: pkgs/applications/misc/jiten/default.nix [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  src = fetchFromGitHub {
+    owner   = "obfusk";
+    repo    = "jiten";
+    rev     = "d98922e8170a82581178577a4c1a85751014c789";   # v0.4.0
+    sha256  = "0l690lfj5rlh9y6aqysh1sgwh13f2yq0vsdq4dimxxpjqixfk20x";
+  };
+  meta = {
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/jiten/default.nix:31]:
```suggestion

  meta = with lib; {
```

File: pkgs/applications/misc/jiten/default.nix [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    '';
+    homepage = "https://github.com/obfusk/jiten";
+    license = with lib.licenses; [
+      agpl3Plus               # code
+      cc-by-sa-30             # jmdict/kanjidic
+      unfreeRedistributable   # pitch data from wadoku is non-commercial :(
+    ];
+    maintainers = [ lib.maintainers.obfusk ];
+    platforms = lib.platforms.linux;
+  };
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/jiten/default.nix:78]:
```suggestion
    license = with licenses; [
      agpl3Plus               # code
      cc-by-sa-30             # jmdict/kanjidic
      unfreeRedistributable   # pitch data from wadoku is non-commercial :(
    ];
    maintainers = [ maintainers.obfusk ];
    platforms = platforms.linux;
  };
```

File: pkgs/applications/misc/jiten/default.nix [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      Command-line interface
+    '';
+    homepage = "https://github.com/obfusk/jiten";
+    license = with licenses; [
+      agpl3Plus               # code
+      cc-by-sa-30             # jmdict/kanjidic
+      unfreeRedistributable   # pitch data from wadoku is non-commercial :(
+    ];
+    maintainers = [ maintainers.obfusk ];
+    platforms = platforms.linux;
```
Comment (obfusk) [inline pkgs/applications/misc/jiten/default.nix:88]: I assume it works on `platforms.unix`, but I've only tested it on linux.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 110900` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>jiten</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

jiten:
<details>

Zero tests run by pytest got detected: 'Ran 0 tests in 0.000s'
You have the following options to fix this:
  - Tell pytest(CheckHook) where to find the tests included in the package
  - Check if the GitHub Repo contains tests but they are not shipped with Pypi. If so please switch to `fetchFromGitHub`.
  - If the Packages does not contain any tests add `'doCheck = false;'` and a `pythonImportsCheck`.

</details>

File: pkgs/applications/misc/jiten/default.nix [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  src = fetchFromGitHub {
+    owner = "obfusk";
+    repo = "jiten";
+    rev = "v${version}";
+    sha256 = "0l690lfj5rlh9y6aqysh1sgwh13f2yq0vsdq4dimxxpjqixfk20x";
+  };
+
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ pcre sqlite ];
+  propagatedBuildInputs = with python3Packages; [ click flask python3 rlwrap ];
```
Comment (sikmir) [inline pkgs/applications/misc/jiten/default.nix:24]:
```suggestion
  propagatedBuildInputs = with python3Packages; [ click flask rlwrap ];
```

**Comment by sikmir** [comment]
```
result/bin/jiten gui
Traceback (most recent call last):
  File "/nix/store/wkw6fsjasr7jbbrlakxxpbiapa8hws42-python3-3.8.7/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/nix/store/wkw6fsjasr7jbbrlakxxpbiapa8hws42-python3-3.8.7/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/nix/store/dzpp7zxzwlzpamldms8w5h4vnwlm8i69-jiten-0.4.0/lib/python3.8/site-packages/jiten/cli.py", line 564, in <module>
    cli(prog_name = name)
  File "/nix/store/aph3vafycfbj9325vgcm7y1zlx5v88j6-python3.8-click-7.1.2/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/nix/store/aph3vafycfbj9325vgcm7y1zlx5v88j6-python3.8-click-7.1.2/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/nix/store/aph3vafycfbj9325vgcm7y1zlx5v88j6-python3.8-click-7.1.2/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/nix/store/aph3vafycfbj9325vgcm7y1zlx5v88j6-python3.8-click-7.1.2/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/nix/store/aph3vafycfbj9325vgcm7y1zlx5v88j6-python3.8-click-7.1.2/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/nix/store/aph3vafycfbj9325vgcm7y1zlx5v88j6-python3.8-click-7.1.2/lib/python3.8/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/nix/store/dzpp7zxzwlzpamldms8w5h4vnwlm8i69-jiten-0.4.0/lib/python3.8/site-packages/jiten/cli.py", line 546, in gui
    start(link)
  File "/nix/store/dzpp7zxzwlzpamldms8w5h4vnwlm8i69-jiten-0.4.0/lib/python3.8/site-packages/jiten/gui.py", line 34, in start
    import webview
ModuleNotFoundError: No module named 'webview'
```

File: pkgs/applications/misc/jiten/default.nix [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    export JITEN_ANDROID=yes          # build & package *.sqlite3
+  '';
+
+  postPatch = ''
+    patchShebangs bin/jiten
+    sed -i 's!/bin/bash!'"$(command -v bash)"'!' Makefile jiten/res/jmdict/Makefile
+  '';
+
+  postFixup = ''
+    wrapProgram $out/bin/jiten \
```
Comment (sikmir) [inline pkgs/applications/misc/jiten/default.nix:37]: Wrap `jiten-gui` as well:
```
$ cat result/bin/jiten-gui
#!/bin/bash
exec jiten gui "$@"
```

File: pkgs/applications/misc/jiten/default.nix (hunk 6 above)
Comment (sikmir) [reply pkgs/applications/misc/jiten/default.nix:88]: Tested build on darwin.

Comment (sikmir) [inline pkgs/applications/misc/jiten/default.nix:88]:
```suggestion
    platforms = platforms.unix;
```

**Comment by obfusk** [comment]
> ModuleNotFoundError: No module named 'webview'

Because the gui is optional and depends on `pywebview` I've removed the `jiten-gui` script from the package and improved the error message a bit.

**Comment by obfusk** [comment]
@SuperSandro2000 @sikmir thanks for the feedback. I've just released `v1.0.0` and updated the PR accordingly.

**Comment by obfusk** [comment]
I resolved the merge conflict. Is there anything else preventing this from being merged?

File: pkgs/applications/misc/jiten/default.nix [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    export JITEN_FINAL=yes            # build & package *.sqlite3
+  '';
+
+  postPatch = ''
+    sed -i 's!/bin/bash!'"$(command -v bash)"'!' Makefile jiten/res/jmdict/Makefile
+  '';
+
+  checkPhase = "make test";
+
+  postInstall = "rm $out/bin/jiten-gui";  # requires pywebview
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/jiten/default.nix:37]:
```suggestion
  postInstall = ''
    # requires pywebview
    rm $out/bin/jiten-gui
  '';
```

File: pkgs/applications/misc/jiten/default.nix [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+  preBuild = ''
+    export JITEN_VERSION=${version}   # override `git describe`
+    export JITEN_FINAL=yes            # build & package *.sqlite3
+  '';
+
+  postPatch = ''
+    sed -i 's!/bin/bash!'"$(command -v bash)"'!' Makefile jiten/res/jmdict/Makefile
+  '';
+
+  checkPhase = "make test";
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/jiten/default.nix:35]:
```suggestion
  checkTarget = "test";
```

File: pkgs/applications/misc/jiten/default.nix [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+  propagatedBuildInputs = with python3Packages; [ click flask ];
+  checkInputs = [ nodejs ];
+
+  preBuild = ''
+    export JITEN_VERSION=${version}   # override `git describe`
+    export JITEN_FINAL=yes            # build & package *.sqlite3
+  '';
+
+  postPatch = ''
+    sed -i 's!/bin/bash!'"$(command -v bash)"'!' Makefile jiten/res/jmdict/Makefile
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/jiten/default.nix:32]: Can we use substituteInPlace here?

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please remove the merge commit. I cannot squash this PR due to the maintainer commit and it makes the history hard to follow when merged.

**Comment by obfusk** [comment]
> Please remove the merge commit. I cannot squash this PR due to the maintainer commit and it makes the history hard to follow when merged.

Sorry. I knew I should have just done a commad-line rebase instead of using the GitHub interface :sweat_smile:

File: pkgs/applications/misc/jiten/default.nix (hunk 11 above)
Comment (obfusk) [reply pkgs/applications/misc/jiten/default.nix:32]: We can. Thanks!

File: pkgs/applications/misc/jiten/default.nix (hunk 10 above)
Comment (obfusk) [reply pkgs/applications/misc/jiten/default.nix]: I had to revert that b/c it tries to run Python tests instead of the `test` target from the `Makefile`.
